strawberryfields/engine.py

```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+
+        # Run a job asynchronously
+        job = engine.run_async(program, shots=1)
+        job.status  # <JobStatus.QUEUED: 'queued'>
+        job.result  # InvalidJobOperationError
+        # (After some time...)
+        job.refresh()
+        job.status  # <JobStatus.COMPLETE: 'complete'>
+        job.result  # [[0 1 0 2 1 0 0 0]]
+
+    Args:
+        target (str): the target device
+        connection (strawberryfields.engine.Connection): a connection to the remote job
+            execution platform
+    """
+
+    POLLING_INTERVAL_SECONDS = 1
+    VALID_TARGETS = ("chip2",)
+
+    def __init__(self, target: str, connection: Connection = None):
+        if target not in self.VALID_TARGETS:
+            raise ValueError(
+                "Invalid engine target: {} (valid targets: {})".format(
+                    target, self.VALID_TARGETS
+                )
+            )
+        if connection is None:
+            # TODO use `load_config` when implemented
+            config = DEFAULT_CONFIG["api"]
+            connection = Connection(
+                token=config["authentication_token"],
+                host=config["hostname"],
+                port=config["port"],
+                use_ssl=config["use_ssl"],
+            )
+
+        self._target = target
+        self._connection = connection
+
+    @property
+    def target(self) -> str:
+        """The target device used by the engine.
+
+        Returns:
+            str: the target device used by the engine
+        """
+        return self._target
+
+    @property
+    def connection(self) -> Connection:
+        """The connection object used by the engine.
+
+        Returns:
+            strawberryfields.engine.Connection: the connection object used by the engine
+        """
+        return self._connection
+
+    def run(self, program: Program, shots: int = 1) -> Optional[Result]:
+        """Runs a remote job synchronously.
+
+        In the synchronous mode, the engine blocks until the job is completed, failed, or
+        cancelled. If the job completes successfully, the result is returned; if the job
+        fails or is cancelled, ``None`` is returned.
 
         Args:
-            program (Program, Sequence[Program]): quantum programs to run
-            shots (int): number of times the program measurement evaluation is repeated
+            program (strawberryfields.Program): the quantum circuit
+            shots (int): the number of shots for which to run the job
+
+        Returns:
+            [strawberryfields.engine.Result, None]: the job result if successful, and
+                ``None`` otherwise
+        """
+        job = self.run_async(program, shots)
+        try:
+            while True:
+                job.refresh()
+                if job.status == JobStatus.COMPLETE:
+                    return job.result
+                if job.status == JobStatus.FAILED:
+                    raise JobFailedError("The computation failed; please try again.")
```

Comment:
I wonder if we'll be able to provide more information here to the user on the cause of failure (e.g. timeout, etc.), but right now I cannot really think of anything else.

Reply:
This is something that should be probably done via logging (i.e., `log.info(message)`, `log.warning(message)` etc).

Reply:
For that particular case (job status becomes "failed"), I believe the most detail we can provide to the user is something along the lines of "The computation failed due to an internal system error." (There may be more specific causes, e.g. hardware error, that is only available in our internal logs.)

@josh146 Do you mean logging a failure message and not raising an exception at all?